### PR TITLE
replace ctor with linkme

### PR DESCRIPTION
**Pull request by garypen**
[ctor](https://crates.io/crates/ctor) is completely replaced with [linkme](https://crates.io/crates/linkme). 

pros:
 - no code executes before main.
 - simplified plugin registration/management

cons:
 - ~all crates must include `linkme 0.3.6`~
 - increased public footprint by exposing PLUGINS, ~linkme~ etc...

This gives us a more stable foundation for plugin loading/registration and we could (in future, not this PR) add mechanisms to exclude plugins or re-order them or ... The main benefit is in reducing the risk of UB creeping in to the codebase in future via plugin registration.

I believe the change is fundamentally a compatible change. ~, but it may require plugins generated using `scaffold` to be modified (to include the `linkme` crate) so that they continue to compile~

### CODE REVIEW

**Review by BrynCooke — APPROVED** [review]

**Review by bnjjj — APPROVED** [review]

File: apollo-router/src/plugin/mod.rs [hunk 1]
```diff
@@ -299,16 +310,13 @@ macro_rules! register_plugin {
...
-                    format!("{}.{}", $group, $name)
-                };
-
-                $crate::plugin::register_plugin::<$plugin_type>(qualified_name);
-            }
+            use $crate::_private::once_cell::sync::Lazy;
+            use $crate::_private::PluginFactory;
+            use $crate::_private::PLUGINS;
+
+            #[$crate::_private::linkme::distributed_slice(PLUGINS)]
```
Comment (SimonSapin) [inline apollo-router/src/plugin/mod.rs:317]: Although [that feature is undocumented](https://github.com/dtolnay/linkme/blob/0.3.6/impl/src/attr.rs#L4), I’ve tested that this change removes the need for plugin crates to have their own dependency on `linkme`, making this PR a non-breaking change:

```suggestion
            #[$crate::_private::linkme::distributed_slice(PLUGINS)]
            #[linkme(crate = $crate::_private::linkme)]
```

Comment (garypen) [reply apollo-router/src/plugin/mod.rs:317]: Great suggestion! Thank you.

Comment (garypen) [reply apollo-router/src/plugin/mod.rs:317]: done: https://github.com/apollographql/router/pull/2181/commits/cf3b62b33fe674d8cdc307cd2a40f9dc33371dc0

File: NEXT_CHANGELOG.md [hunk 2]
```diff
@@ -26,6 +26,19 @@ By [@USERNAME](https://github.com/USERNAME) in https://github.com/apollographql/
...
+### improve plugin registration predictability ([PR #2181](https://github.com/apollographql/router/pull/2181))
+
+This replaces [ctor](https://crates.io/crates/ctor) with [linkme](https://crates.io/crates/linkme). `ctor` enables rust code to execute before `main`. This can be a source of undefined behaviour and we don't need our code to execute before `main`. `linkme` provides a registration mechanism that is perfect for this use case, so switching to use it makes the router more predictable, simpler to reason about and with a sound basis for future plugin enhancements.
+
+This is marked as a "breaking" change, because any existing rust plugins will need to add:
+
+```
+linkme = "0.3"
+```
+to their list of `Cargo.toml` dependencies. No router interfaces are incompatibly changed by this.
```
Comment (SimonSapin) [inline NEXT_CHANGELOG.md:38]: Needs update, and to be moved to another section

File: apollo-router-scaffold/templates/base/Cargo.toml [hunk 3]
```diff
@@ -27,6 +27,7 @@ apollo-router = "1.4.0"
 {{/if}}
 async-trait = "0.1.52"
 futures = "0.3.21"
+linkme = "0.3.6"
```
Comment (SimonSapin) [inline apollo-router-scaffold/templates/base/Cargo.toml:30]: This can also be removed

**Review by SimonSapin — APPROVED** [review]
(Note that ctor is still in the dependency graph, through `proteus` / `typetag` / `inventory`)

File: apollo-router-scaffold/templates/base/Cargo.toml (hunk 3 above)
Comment (garypen) [reply apollo-router-scaffold/templates/base/Cargo.toml:30]: Yeah. I did think about that. I figured that leaving it in there won't cause an issue, but I can remove it to be consistent.

File: NEXT_CHANGELOG.md (hunk 2 above)
Comment (garypen) [reply NEXT_CHANGELOG.md:38]: > (Note that ctor is still in the dependency graph, through `proteus` / `typetag` / `inventory`)

 It's still better to not use `ctor` if you don't need it, so steps in the right direction...

File: apollo-router-scaffold/templates/base/Cargo.toml (hunk 3 above)
Comment (garypen) [reply apollo-router-scaffold/templates/base/Cargo.toml]: done: https://github.com/apollographql/router/pull/2181/commits/57c557e1705707cd57c99097efa0dc13686817a0

File: NEXT_CHANGELOG.md (hunk 2 above)
Comment (garypen) [reply NEXT_CHANGELOG.md]: done: https://github.com/apollographql/router/pull/2181/commits/57c557e1705707cd57c99097efa0dc13686817a0

**Review by SimonSapin — APPROVED** [review]

**Pull request merged**